### PR TITLE
`ZipCycle` and `ZipCycleBiased`

### PR DESCRIPTION
**Pull request by sheganinans**
Follow up from: #326

### CODE REVIEW

**Comment by sheganinans** [comment]
Ah, seems to break on Travis. Seems like I need to figure out how to replicate their env.

File: src/lib.rs [hunk 1]
```diff
@@ -389,6 +393,60 @@ pub trait Itertools : Iterator {
...
+    /// let it = (1..5).zip_cycle_biased(0..2);
+    /// assert_eq!(it.unwrap().collect::<Vec<_>>(), vec![(1, 0), (2, 1), (3, 0), (4, 1)]);
+    /// let it2 = (0..2).zip_cycle_biased(1..5);
+    /// assert_eq!(it2.unwrap().collect::<Vec<_>>(), vec![(0, 1), (1, 2)]);
+    /// ```
+    fn zip_cycle_biased<J>(self, other: J) -> Result<ZipCycleBiased<Self, J::IntoIter>, ZipCycleBiasedErr>
+    where
+        J: IntoIterator,
+        <J as IntoIterator>::IntoIter: Clone,
+        Self: Sized + Clone,
```
Comment (jplatte) [inline src/lib.rs:445]: This `Clone` only seems necessary for the unbiased version.

File: src/zip_cycle.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        size_hint::max(self.a.size_hint(), self.b.size_hint())
+    }
+}
+
+// DoubleEndedIter todo.
+
+impl<A, B> ExactSizeIterator for ZipCycle<A, B>
+    where A: ExactSizeIterator + Clone,
+          B: ExactSizeIterator + Clone
+{}
```
Comment (jplatte) [inline src/zip_cycle.rs:119]: Missing final newline.

**Review by jplatte — COMMENTED** [review]
I find the names for these unintuitive. I could never guess what `zip_cycle_biased` does based on its name. Maybe a freestanding `cycling_zip` for the non-biased one and `.zip_cycle_biased` => `.zip_cycle`? Not sure.

File: src/zip_cycle.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    /// The case the right hand side is empty
+    SecondIterEmpty
+}
+
+pub fn zip_cycle<A, B>(a: A, b: B) -> Result<ZipCycle<A, B>, ZipCycleErr>
+where
+    A: Clone + Iterator,
+    B: Clone + Iterator,
+{
+    match (a.clone().peekable().peek(), b.clone().peekable().peek()) {
```
Comment (phimuemue) [inline src/zip_cycle.rs:45]: Is it really necessary to make a `peekable` here? Couldn't we just check `a.clone().next()`?

File: src/zip_cycle_biased.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+pub enum ZipCycleBiasedErr {
+    /// The error result for when the Cycling half of the iter is empty
+    CycleIterEmpty
+}
+
+pub fn zip_cycle_biased<I, C>(i: I, c: C) -> Result<ZipCycleBiased<I, C>, ZipCycleBiasedErr>
+where I: Iterator,
+      C: Clone + Iterator
+{
+    match c.clone().peekable().peek() {
```
Comment (phimuemue) [inline src/zip_cycle_biased.rs:33]: Similarly here: Wouldn't `c.clone().next()` be enough?

**Comment by Philippe-Cholet** [comment]
This seems abandoned so I'm closing this. I would promptly reopen it if needed.

**Pull request closed without merging**